lib/IDE/CodeCompletion.cpp

```diff
@@ -3474,8 +3482,20 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
     // operator, filling in the operator and dummy right-hand side.
     sequence.push_back(nullptr); // operator
     sequence.push_back(nullptr); // RHS
-    auto *SE = SequenceExpr::create(CurrDeclContext->getASTContext(), sequence);
-    prepareForRetypechecking(SE);
+    for (unsigned i = 0; i < sequence.size(); i++) {
+      if (sequence[i]) {
+        sequence[i] = prepareForRetypechecking(sequence[i]);
+      }
+    }
+
+    // Set up a new ASTContext in which we can allocate SequenceExpr nodes that
+    // we try and typecheck. After trying all operators, we can delete it with
+    // all the temporary SequenceExpr within it.
+    auto &CurrCtxt = CurrDeclContext->getASTContext();
+    auto ScratchCtxt =
+        ASTContext::get(CurrCtxt.LangOpts, CurrCtxt.SearchPathOpts,
+                        CurrCtxt.SourceMgr, CurrCtxt.Diags);
+    SWIFT_DEFER { delete ScratchCtxt; };
```

Comment:
If I understand correctly, you are creating a throwaway ASTContext just to allocate a SequenceExpr outside the main ASTContext. Generally I think:

* Releasing the SequenceExpr allocated memory separately from the main ASTContext doesn't matter much. Completion requests do not persist any memory (the main ASTContext goes away after the request is done) so it doesn't make much difference if the memory gets allocated via the main ASTContext or another one that gets released sooner.
* This is **very** risky, you introduce chance that, after the codebase changes, the SequenceExpr could leak out later and have a dangling pointer around, or someone makes more use of this alternate `ASTContext` without paying enough attention of the lifetime of the created objects or how it is used, ending up with hard to find and reproduce memory bugs.

If there's sufficient need to allocated AST nodes in a separate allocation arena we should think on how to provide specific APIs for this, and what the ramifications are; I don't think it is a good idea to create a separate ASTContext for this use here.